// @progress/kendo-react-data-tools/dist/npm/navigation/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _a, _b, _c;
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_LEVEL = 'data-keyboardnavlevel';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_SCOPE = 'data-keyboardnavscope';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_HEADER = 'data-keyboardnavheader';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_BODY = 'data-keyboardnavbody';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_ID = 'data-keyboardnavid';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_DATA_ZONE = 'data-keyboardnavzone';
/**
 * @hidden
 */
exports.KEYBOARD_NAV_FILTER_COL_SUFFIX = '_filter';
var FOCUSABLE_ELEMENTS_BASE = [
    'input:not([disabled]):not([type=hidden])',
    'select:not([disabled])',
    'textarea:not([disabled])',
    'button:not([disabled])',
    'a[href]',
    'area[href]',
    'summary',
    'iframe',
    'object',
    'embed',
    'audio[controls]',
    'video[controls]',
    '[contenteditable]'
];
/**
 * @hidden
 */
exports.FOCUSABLE_ELEMENTS = FOCUSABLE_ELEMENTS_BASE.concat(['[tabindex]']);
/**
 * @hidden
 */
exports.TABBABLE_ELEMENTS = FOCUSABLE_ELEMENTS_BASE.concat(['[tabindex]']).map(function (selector) { return selector + ':not([tabindex="-1"])'; });
/**
 * @hidden
 */
exports.tableKeyboardNavigationScopeAttributes = (_a = {},
    _a[exports.KEYBOARD_NAV_DATA_SCOPE] = true,
    _a);
/**
 * @hidden
 */
exports.tableKeyboardNavigationHeaderAttributes = (_b = {},
    _b[exports.KEYBOARD_NAV_DATA_HEADER] = true,
    _b);
/**
 * @hidden
 */
exports.tableKeyboardNavigationBodyAttributes = (_c = {},
    _c[exports.KEYBOARD_NAV_DATA_BODY] = true,
    _c);
